[#2776] Rate limiter for request calls to node server

LetItRock opened the pull request:
<!-- FILL OUT THE BELOW FORM OR YOUR PR WILL BE AUTOMATICALLY CLOSED -->
**Status**
- [x] WIP
- [ ] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api
- hyperion (frontend)

**Release notes for users (delete if codebase-only change)**
- express middleware which will limit request calls to API for authenticated/anonymous user within a time frame; it will create a key in redis "rateLimiter:limit:id"; where id for authenticated user is his "id" and for anonymous user "ip address";

We should also consider values for max requests and for time frame.
Do we want to implement same middleware for other servers? Which one?

mxstbr reviewed (commented):
This looks amazing! 😍 We definitely also want to use this in hyperion, although hyperion probably needs stricter limits than the API. (since it's rendering requests take a while longer than the API responses)

I'm not very familiar with rate limiting, so I'm not 100% sure what the limiting values should be. Any ideas?

LetItRock commented:
Ok, I'll also implement this in hiperion. Me neither, maybe we should look into statistics/analytics?

mxstbr commented:
Yeah, maybe have a look around/google what other people are doing, I'm really not sure.

brianlovin committed:
Merge pull request #2758 from withspectrum/stop-thread-spam

Stop thread spam

StefanNemeth committed:
Re-add mp4 as an accepted media input

brianlovin committed:
Merge pull request #2778 from StefanNemeth/alpha

Paste image from clipboard into chat input

brianlovin committed:
Fix opacity on media preview

brianlovin committed:
Merge pull request #2797 from withspectrum/fix-image-preview-opacity

Fix opacity on media preview

brianlovin committed:
Potential fix

brianlovin committed:
Fix thread view auth links

brianlovin committed:
One more check

brianlovin committed:
Merge pull request #2799 from withspectrum/fix-messages-tab

Fix messages tab

brianlovin committed:
Merge pull request #2798 from withspectrum/fix-thread-view-auth

Fix thread view auth links

brianlovin committed:
Swap out sentry dsn to secret ones

brianlovin committed:
Fixes user profiles in top members component of analytics

brianlovin committed:
Merge pull request #2709 from withspectrum/restrict-deployments-to-sfo

Restrict deployments to SFO

LetItRock committed:
Implement rate limiter for hyperion;

LetItRock commented:
I've implemented this also for hyperion ;)

brianlovin committed:
Better email failure handling in hermes

brianlovin committed:
Remove console

superbryntendo committed:
addresses brian's feedback. adds documentation for the behavior of granularUserProfile.

superbryntendo committed:
flow errors

mxstbr committed:
Merge pull request #2802 from withspectrum/fix-rep-in-analytics

Fixes user profiles in top members component of analytics

Tom-Bonnike committed:
Fix <main> wrapper causing layout issues

mxstbr committed:
Merge branch 'alpha' into fix-tests

mxstbr committed:
Fix some more bugs

mxstbr committed:
Further fixes

mxstbr committed:
exit process on unhealthy db in test env

mxstbr committed:
Fix top communities pure

datio committed:
Update blacklist with routes currently in use

StefanNemeth committed:
Fix syntax of media input accept string

StefanNemeth committed:
Set blob to null instead of creating new instance

StefanNemeth committed:
Flow fix

mxstbr committed:
Fix mobile cold start

mxstbr committed:
Merge pull request #2809 from withspectrum/fix-mobile-cold-start

Fix mobile cold start

brianlovin committed:
Swap spectrum api keys in alpha to keep in sync with prod hotfix

brianlovin committed:
Merge pull request #2812 from withspectrum/swap-spectrum-api-key

Swap spectrum api keys in alpha to keep in sync with prod hotfix

brianlovin commented:
@mxstbr are you cool owning the testing on this one and deploying to alpha?

brianlovin committed:
Fix flow

brianlovin committed:
Merge pull request #2808 from StefanNemeth/fix/blob-undefined

Fix "Blob undefined"

brianlovin committed:
Review fixes

brianlovin committed:
Merge branch 'fix-hermes-failures' of github.com:withspectrum/spectrum into fix-hermes-failures

brianlovin committed:
Merge pull request #2752 from Tom-Bonnike/header-a11y

Improve NavBar’s a11y

brianlovin committed:
Merge pull request #2674 from withspectrum/fix-tests

Fix e2e tests

brianlovin committed:
Merge branch 'alpha' into sentry-secrets

brianlovin committed:
Merge pull request #2801 from withspectrum/sentry-secrets

Swap out sentry dsn to secret ones

brianlovin committed:
Sort and preserve order of participant heads

brianlovin committed:
Remove one more console error

brianlovin committed:
Merge pull request #2806 from withspectrum/fix-hermes-failures

Better email failure handling in hermes

superbryntendo committed:
gets rid of extraneous main element and moves it to replace individual wrappers. I think long-term we can come up with a better, more generalized solution, but this seems to fix the issues...

superbryntendo committed:
fixes firefox's breaking dm layout issues.

superbryntendo committed:
reverts a bunch of changes made when the main was added at the top level which didn't work out.

brianlovin committed:
Add flow typings

LetItRock committed:
Create UsernameSearch component which holds validation for username; use it in two places UserSettings and SetUsername;

mxstbr committed:
Merge pull request #2816 from withspectrum/fix-participant-heads-reordering

Sort and preserve order of participant heads

mxstbr committed:
Clean up flow types

mxstbr reviewed (commented):
I looked around, and [Trello allows 300 requests per 10 seconds to their API](https://help.trello.com/article/838-api-rate-limits). (which is an avg of 30 requests per second)

That seems very high to me—our frontend when used by a legitimate user does at most probably like 40-50 requests in a 10 second window. (and that's when clicking around like a mad man and not doing anything actually useful)

Let's try restricting the API to 100 req per 10 sec? (avg of 10 req per second, still a ton)

----

Hyperion will need a way lower limit because it's a lot more expensive to render a page. Thinking out loud about this, no normal user should ever have to make more than 5 req/sec, so let's try restricting hyperion to 50 requests in 10 sec?

chrbala commented:
Make sure to look into the VPN case if you're grouping by IP. It's possible that there might be a lot of users on the same IP if they're all on the same VPN. 

We occasionally get rate limited on sites because of the number of users on our corporate network.

mxstbr committed:
Merge pull request #2783 from LetItRock/set-username-validation-fixes

Better validation for username and logic behind the calls to API in SetUsername component

mxstbr reviewed (commented):
We should make sure that hyperion still serves cached responses for logged-out users, and only rate limit new renders.

datio committed:
Update blacklist with routes currently in use

#2807

mxstbr committed:
Merge pull request #2807 from datio/patch-1

Update blacklist with routes currently in use

mxstbr committed:
Fix token-based auth for mobile apps

mxstbr committed:
Only allow explicit listens to thread updates

mxstbr committed:
Merge pull request #2823 from withspectrum/fix-token-auth

Fix token-based auth for mobile apps

johnnynotsolucky committed:
Use the closeModal handler for Cancel buttons

mxstbr committed:
Merge pull request #2824 from johnnynotsolucky/close-modal

Implement close modal handler for create/restore channel modals

mxstbr committed:
Update draft-js-markdown-plugin

This fixes the annoying bug where creating headings or lists would move
your cursor above the heading/list.

LetItRock commented:
@brianlovin I'm not added to "now" team. I can try to own testing and deploying to alpha if I will have permissions to do that.

@mxstbr I think the same way as you that 100 its a lot, so lets restrict API calls to 100 per 10 sec for now. 
I was testing hyperion and figured out that 1 page refresh for ex. when I'm on home page makes 7 calls to RateLimiter... is that related to resource downloading?

@chrbala good catch, could you point me how can we archive this?

mxstbr commented:
> I was testing hyperion and figured out that 1 page refresh for ex. when I'm on home page makes 7 calls to RateLimiter... is that related to resource downloading?

Arg I totally forgot about the static files, damn. Can we add the rate limiting middleware after the `express.static` one?

brianlovin committed:
Merge pull request #2826 from withspectrum/update-markdown-plugin

Update draft-js-markdown-plugin

LetItRock commented:
@mxstbr sure... and I will check that later today ;)

andreasklinger commented on api/routes/middlewares/index.js:
```diff
@@ -50,6 +50,10 @@ import passport from 'passport';
 
+// rate limiter middleware to reduce API calls to server within time frame
+import rateLimiter from 'shared/middlewares/rateLimiter';
+middlewares.use(rateLimiter('api', 2500, 3600000));
```
outsider pov: 
imo bad (aka magic) numbers b/c i need to look at source what they mean 

`rateLimiter('api', MAX_REQUESTS, ONE_HOUR)`

normally i'd say only name constants and variables by WHAT not HOW but i think stuff like ONE_HOUR is an exception b/c you would do it with syntax like `1.hour` if you could - so it's more a workaround to make it readable in JS

mxstbr replied:
Yeah totally, I don't think @LetItRock had thought those numbers through yet

LetItRock replied:
@andreasklinger yeah for sure I will extract those values to more readable constants... till today I didn't know what exact values they gonna be... and we are still debating on values for hyperion...

brianlovin committed:
Migration and protections against lowercase usernames

brianlovin committed:
Clean up all channel and community slugs as well, consolidate lowercase logic to db

brianlovin committed:
Handle loader mismatches

superbryntendo committed:
fixes navbar positioning on DM and Notifications views

superbryntendo committed:
Merge branch 'grid-conversion' of github.com:withspectrum/spectrum into grid-conversion

brianlovin committed:
Small tweak to migration

superbryntendo committed:
fixes a case where the input gets pushed offscreen in DM threads

superbryntendo committed:
managing height against the tab bar is not ideal. v excited to convert views to grid/subgrid to manage this kind of thing...

superbryntendo committed:
semi-colons are fun

brianlovin committed:
Merge pull request #2818 from withspectrum/grid-conversion

Fixes all the <main> changes by replacing the wrappers with <main> vs adding an extra element

brianlovin committed:
Clean up and refactor user settings view to align with other settings

brianlovin committed:
Fix annoying redirect on save

brianlovin committed:
Set up initial mutation

brianlovin committed:
Implement client side component for deleting account

superbryntendo committed:
added actions section for handling additional contextual actions.

brianlovin committed:
Implement deleting user mutation

superbryntendo committed:
forgot to remove a dumb experiment

brianlovin committed:
Styling and frontend cleanup, prevent deleting if is pro

brianlovin committed:
Debugging deletion while is pro

LetItRock commented:
@mxstbr I've moved rate limiter middleware code below "cache"... and we now get few requests less to rate limiter... those are:
- / 
- /static/js/bootstrap.js
- /undefined/img/favicon.ico

I think that it should be possible to reduce those two last right? :P 
......................

Another thing is VPN/Proxy, as far as I know they basically almost the same thing (VPN encrypts data - proxy not)... I know that proxies are using "X-Forwarded-For" header to kind of "store" real user IP address... but I'm not sure about VPN do they handle requests same way with same header? Can someone confirm that?
Our RateLimiter middleware is using request-ip library which handles "X-Forwarded-For" and other specific headers from [docs](https://github.com/pbojinov/request-ip)

mxstbr committed:
Merge branch 'alpha' into better-code-blocks-in-messages

mxstbr committed:
WIP: Make typing work without breaking anything else

mxstbr committed:
Convert text message to DraftJS in addMessage mutation

mxstbr committed:
Chat input send plain text if on Android

mxstbr committed:
Fix Android browser detection

mxstbr committed:
Fix styling of fallback plaintext input

mxstbr committed:
Move conversion from text to draftjs earlier

mxstbr commented:
> - /static/js/bootstrap.js
> - /undefined/img/favicon.ico

What on earth is that second one?! The first one should be served with `build/static/js/bootstrap.js` it's interested that the static middleware doesn't see that one...

mxstbr commented:
Oh actually, that first one should be fixed in `src/views/navbar/index.js` and `src/views/pages/components/nav.js`—it says `href={${process.env.PUBLIC_URL}/img/favicon.ico}` in both those places, but PUBLIC_URL is no longer defined so let's try and change that to `/img/favicon.ico`.

mxstbr committed:
Merge pull request #2781 from withspectrum/profile-list-visibility

adds membership lists to the left profile column on mid-size displays

mxstbr committed:
Restart API in CI if it crashes during Cypress tests

I thought I implemented that behavior, but as it turns out nodemon
will actually only restart the server if a file changes.

Let's see if this fixes our integration test suite!

mxstbr committed:
Make production API be usable for tests

mxstbr committed:
Run production API on Circle

mxstbr committed:
Fix circleci config

mxstbr committed:
Use forever to restart API during testing

mxstbr committed:
Update testing docs

mxstbr committed:
Wait for servers to start before running test suite

brianlovin committed:
Clean up pending and blocked users styling and layout

brianlovin committed:
Small consistency fixes

mxstbr committed:
Comment out failing tests

mxstbr committed:
Waiting for RethinkDB doesnt work

mxstbr committed:
Merge pull request #2836 from withspectrum/fix-android-once-and-for-all

Work around DraftJS on Android in the chat input

mxstbr committed:
Try and move rethinkdb to docker images

mxstbr committed:
Make the API return something for the root route

mxstbr committed:
Wait simulataneously

mxstbr committed:
Remove console.logs

mxstbr committed:
Merge pull request #2839 from withspectrum/wait-for-servers-to-start-in-ci

Fix integration tests and refactor CI to properly wait for servers to be started

brianlovin committed:
Merge branch 'alpha' of github.com:withspectrum/spectrum into fix-multi-actions

brianlovin committed:
Merge pull request #2840 from withspectrum/fix-multi-actions

Clean up pending and blocked users styling and layout

dev-drprasad committed:
:recycle: Catch queue failed events and report to sentry

Ponjimon committed:
Fix the deprecation warning of Draft.js

mxstbr committed:
Merge pull request #2845 from lookapanda/draftentity-deprecation

Fix the deprecation warning of Draft.js

LetItRock commented:
@mxstbr 
So the issue with favicon is fixed... for the bootstrap requests I see strange thing is happening... it basically returns index.html page for /static/js/bootstrap.js request 😄 thats why it goes through RateLimiter... 
I've checked package.json and I don't see any dependency to bootstrap... but in config-overrides we have CommonChunksPlugin which creates the bootstrap.js chunk for us... but I've checked that file from served file and it only has webpack config inside... no bootstrap code...

---
@mxstbr @brianlovin And I need someone to answer my question regarding to VPN like few messages above...

brianlovin committed:
Merge branch 'alpha' of github.com:withspectrum/spectrum into fix-username-capitalization-in-urls

brianlovin committed:
Implement custom scalar for lowercase strings

brianlovin committed:
Swap strings and emails to LowercaseStrings

brianlovin committed:
Merge branch 'alpha' of github.com:withspectrum/spectrum into delete-users

brianlovin committed:
Fix pro stripe config, ensure delete logic if pro

brianlovin committed:
Set up FAQ page

brianlovin committed:
eslint cleanup

brianlovin committed:
Add integration test for FAQ page

brianlovin committed:
Add more fields to be removed on user deletion

brianlovin committed:
Add 'deleted' name to show in the ui

brianlovin committed:
First pass at FAQ content

brianlovin committed:
Disable product navbar on faq page

brianlovin committed:
Update docs

brianlovin committed:
Add integration test for deleting users

brianlovin committed:
Fix seeding data with default usersSettings

brianlovin committed:
Fix tests

mknepprath committed:
Add pointer when hovering over message images

mxstbr committed:
Fix logging into mobile for new contributors

Without the `@api-token-secret` in the `now-secrets.json` file the
token-based authentication doesn't work-which means you cannot log into
mobile.

mxstbr committed:
Let anybody log into Expo locally

mxstbr committed:
Merge pull request #2849 from withspectrum/add-api-token-secret

Fix local mobile singin

mxstbr committed:
Run testing setup script before starting API in test mode

Closes #2810

brianlovin committed:
Add styling for code blocks

brianlovin committed:
Fixes login modal grid on desktop

mxstbr committed:
Merge pull request #2855 from withspectrum/fix-code-blocks

Add styling for code blocks

mxstbr committed:
Merge pull request #2850 from withspectrum/fix-initial-integration-test-stup

Run testing setup script before starting API in test mode

brianlovin committed:
Merge pull request #2848 from mknepprath/message-image-cursor

Add pointer when hovering over message images

brianlovin committed:
Merge pull request #2856 from withspectrum/fix-login-modal-width

Fixes login modal grid on desktop

alejandronanez committed:
remove linter errors

mxstbr committed:
Merge branch 'catch-queue-failure-events' of git://github.com/dev-drprasad/spectrum into dev-drprasad-catch-queue-failure-events

mxstbr commented:
- I'm not sure about the VPN thing. Would you mind googling around for the best way to solve that?
- With `bootstrap.js`, if you run `yarn run build:web` and then check the `build/static/js` folder you'll see that it contains the `bootstrap.js` file. It's weird that it would not be served though?

mxstbr committed:
Empty commit to force CircleCi rebuild

mxstbr committed:
Merge pull request #2844 from dev-drprasad/catch-queue-failure-events

♻️ Catch queue failed events and report to sentry

mxstbr committed:
Merge pull request #2857 from alejandronanez/linter-cleanup

remove linter warnings

LetItRock commented:
@mxstbr ok I will try to find something...

---
@mxstbr I've also checked that bootstrap file and it only has webpack specific code, it doesn't have anything related to bootstrap... maybe its some leftover... not sure... because I've tried to remove that chunk config from config-overrides and it was working fine for local development config... 

but on hyperion side I saw something strange happening first application was rendered normally and then after few second it disappeared... have no idea why its happening...
Also I got after those changes new request in RateLimiter for "/static/js/Splash.chunk.js"... weird... Its somehow related to routes.js file...

brianlovin committed:
Reserve right to manage communities immediately

brianlovin committed:
Fix nits

brianlovin committed:
Merge pull request #2833 from withspectrum/fix-username-capitalization-in-urls

Smarter lowercase checks and storage for users, communities, and channels

tomscholz committed:
Fix broken deployment link

mxstbr committed:
Merge pull request #2860 from tomscholz/patch-1

Fix broken deployment link

LetItRock committed:
Fix for bug with /favicon.ico;
RateLimiter extracted magic numbers to constants;
Added debug logs;

mxstbr committed:
Update draft-js-markdown-plugin

- Fixes pesky "Jump above newly created block"-bug
- Adds code language select to thread editor

mxstbr committed:
Cleanup

ArturKlajnerok committed:
Removed svg default avatars from mobile app

ArturKlajnerok committed:
Splash view content should take full width

ArturKlajnerok committed:
CenteredView for loading text on ThreadFeed

mxstbr committed:
Cache homescreen icons from the ServiceWorker

We had to disable caching for all images as Zeit was rate-limiting us
due to too many requests being sent from a single browser in a short
period of time.

Unfortunately, that broke adding-to-homescreen as the icons wouldn't be
sent properly. This patch fixes it by caching homescreen icons in the
ServiceWorker. Closes #2822

mxstbr committed:
Maybe fix DM badge not showing up

This might or might not fix the DM badge not showing up—it's more solid
logic anyway.

mxstbr committed:
Remove console.log

mxstbr committed:
draftjs-editor => rich-text-editor

mxstbr committed:
Create android-friendly wrapper around draft-js-plugins-editor

Instead of making the Android fix specific to the chat input this patch
moves the Android fix a level deeper inside of a custom wrapper
component around draft-js-plugins-editor.

This custom wrapper component conditionally renders either the standard
draft-js-plugins-editor, or a vanilla text input if on Android. This
should theoretically allow us to reuse the same fix for the main thread
composer so folks can also create threads on Android.

ArturKlajnerok committed:
Simplified Avatar and removed unneded code in native app

ArturKlajnerok committed:
Updated FacepileContainer to use latest Avatar component in native app

ArturKlajnerok committed:
Updated ThreadItem to use latest Facepile component and added few layout changes

ArturKlajnerok committed:
Styling ThreadItem in native app

LetItRock commented:
@mxstbr finally I got it how it works on hyperion side: every route change for ex. "/login", "/about" renders page on server side and then caches it in redis if user is not authenticated... 
Splash.chunk.js thats basically Pages component which renders different components based on routes... 
RateLimiter in hyperion will count every route request and "*.chunk.js" files for authenticated users... it will associate "user id" with number of possible requests in time frame... 
but for those who are not authenticated it will count only first time (by IP) and it will render that page and save in cache... every other request comming from unauthenticated users will be taken from cache... thats in hyperion case...

VPN - it looks like its not possible to access "real user IP" because VPN its like separate private network on your computer and all the requests are going through VPN server which has one IP address for all users connected to that private network... so on our end we will se VPN server IP address... thats my understanding...

Github has implemented rate limiting for API in second way: 5000 requests for authenticated user per hour, 60 requests for unauthenticated users per hour (associated with IP) [link](https://developer.github.com/v3/#rate-limiting);

StackExchange: 30 requests per second for application (associated by IP), if number is higher then they give ban for 30 sec... they also have daily limit for application without access_token to 10000, this number is given for all applications from same IP address... if the application has access_token that means its associated with user and it will have its app/user limit 10000 requests per day [link](https://api.stackexchange.com/docs/throttle);

Twitter: only limits authenticated users/applications and they have different limits to different endpoints [link](https://developer.twitter.com/en/docs/basics/rate-limiting), [link](https://developer.twitter.com/en/docs/basics/rate-limits);

SoundCloud: they calculate rate limits based on clients application client id and its 15000 per day [link](https://developers.soundcloud.com/docs/api/rate-limits);

Facebook [link](https://developers.facebook.com/docs/graph-api/advanced/rate-limiting) and Youtube [link](https://developers.google.com/youtube/v3/getting-started#quota) they have more compicated rate limits

spartDev committed:
shows a cheat sheet of supported markdown features

close #2864

brianlovin committed:
Request screenshots in pull request template

spartDev committed:
remove unsuported ~strike~

brianlovin committed:
Merge branch 'alpha' into delete-users

spartDev committed:
fix markdownHint for bold and italics

spartDev committed:
remove underscore syntax

brianlovin committed:
Try to fix delete user test, runs locally fine

mxstbr committed:
Merge pull request #2868 from withspectrum/draft-js-fixies

Create android-friendly wrapper around draft-js-plugins-editor

mxstbr committed:
Merge pull request #2866 from withspectrum/maybe-fix-dm-badge

Maybe fix DM badge not showing up

mxstbr committed:
Merge pull request #2865 from withspectrum/cache-icons-in-sw

Cache homescreen icons from the ServiceWorker

mxstbr committed:
Merge pull request #2847 from withspectrum/delete-users

Delete users

mxstbr committed:
Merge branch 'alpha' into update-markdown-plugin-to-v1.4.3

mxstbr committed:
Fix missing file

mxstbr committed:
Merge pull request #2863 from withspectrum/update-markdown-plugin-to-v1.4.3

Update draft-js-markdown-plugin

spartDev committed:
add cypress test

ArturKlajnerok committed:
Using shared UserInfoType in mobile thread

mxstbr commented:
Thanks for that in-depth research @LetItRock! Thoughts:

- Given that cached requests don't count towards the rate limit, and that every unauthenticated request to hyperion will respond with a cached request after the first time, it seems fine to keep that limit rather low even with VPNs since they'll get cached responses anyway.
- API queries really aren't as expensive to resolve, in comparison to the rendering that hyperion does, so we can keep a lenient limit there so even unauthenticated users behind VPNs can browse the site.

With that in mind, what about the following limits:

- **API: 300 reqs/10 secs**—very lenient, no legitimate user even behind a VPN should ever crack that limit, but it still protects the API from too many requests
- **Hyperion: 50 reqs/20 secs**—there's no way any legitimate user ever opens 50 uncached pages and has hyperion render them in a 20 second window, not even behind a VPN, given that any page you open after the initial load will be served with the app shell locally.

mxstbr committed:
Fix integration test

mxstbr committed:
Remove redundant integration test

mxstbr committed:
Get rid of toPlainText usage to make things faster

mxstbr committed:
Use .hasText instead of .getLength

ArturKlajnerok committed:
Disabeling CI Danger run for forks

mxstbr committed:
Merge pull request #2875 from ArturKlajnerok/danger-fork

Disabeling CI Danger run for forks

mxstbr committed:
Fix fallback Android input in DMs

Eric013 committed:
Update README.md

Fix Testing link

ArturKlajnerok committed:
Fixed e2e id name in channelList

brianlovin committed:
Styling and eslint

brianlovin committed:
Merge pull request #2879 from Eric013/patch-1

Update README.md

brianlovin committed:
Merge pull request #2876 from withspectrum/fix-android-fallback-input-on-dms

Fix fallback Android input in DMs

brianlovin committed:
Update PULL_REQUEST_TEMPLATE.md

brianlovin committed:
Merge pull request #2870 from spartDev/markdown-hints

shows a cheat sheet of supported markdown features

mxstbr committed:
Merge pull request #2871 from withspectrum/update-pr-template

Request screenshots in pull request template

Ponjimon committed:
Fix wrong argument for clickShareLink

ArturKlajnerok committed:
Support FORCE_DEV in stripe config

ArturKlajnerok committed:
Create now-secret.json for CI test build

brianlovin committed:
Fix packages for deploying workers

brianlovin committed:
Add pluto as deploy option in PR template

brianlovin committed:
Eslint

LetItRock commented:
Yeah those thoughts are more clear 😃 thanks 👍 
I think that those limits should be enough even for unauthenticated users through VPN...
we need to give it a try and then will adjust it on the need... do we have some analytics tool?

Maybe it would be good to hear opinion from someone else from team on this? 

I will change those limits after we will have final decision...

brianlovin committed:
Fix infinite scroll loading behavior on tall screens

superbryntendo committed:
returns the gradients to the text bubbles

mxstbr committed:
Merge pull request #2891 from withspectrum/bubble-fix

returns the gradients to the text bubbles

mxstbr committed:
Merge pull request #2888 from withspectrum/fix-worker-deploys

Fix packages for deploying workers

mxstbr committed:
Merge pull request #2889 from withspectrum/pluto-deploy-option

Add pluto as deploy option in PR template

mxstbr committed:
Merge pull request #2881 from ArturKlajnerok/fixing-test

Fixing tests

mxstbr commented:
> we need to give it a try and then will adjust it on the need... do we have some analytics tool?

It'd be great to log to Sentry everytime somebody exceeds the rate limit with their IP or user ID and what they were requesting. Let's get those changes in and ship it!

mxstbr committed:
Merge pull request #2883 from lookapanda/upsell-fix

Fix wrong argument for clickShareLink

mxstbr committed:
Merge pull request #2869 from ArturKlajnerok/alpha

[Mobile v1] Thread Card improvements & styling

mxstbr committed:
Update mobile dependencies

This updates us to Expo SDK v26, and with that React v16.3 and
react-navigation v1.5.8. I also updated `exp` (the expo dev CLI) while I
was at it

mxstbr committed:
Merge pull request #2893 from withspectrum/update-mobile-deps

Update mobile dependencies

brianlovin committed:
Update banning users docs

mattgaunt committed:
Use the users username for the `return to profile` link

brianlovin committed:
Merge pull request #2902 from mattgaunt/fix-return-link

Use the users username for the `return to profile` link

brianlovin committed:
Add forked infinitescroll component

brianlovin committed:
Add check for tallscreens

brianlovin committed:
Remove unused helper

brianlovin committed:
Implement new infinitescroller into views

brianlovin committed:
Remove console

brianlovin committed:
Fix flow

brianlovin committed:
Fix flow again

brianlovin committed:
Add configs to cypress and circleci

mxstbr committed:
Merge pull request #2899 from withspectrum/update-banning-users-docs

Update banning users docs

brianlovin committed:
Merge pull request #2903 from withspectrum/add-dashboard-to-cypress

Add configs to cypress and circleci

brianlovin committed:
Merge branch 'alpha' of github.com:withspectrum/spectrum into pagination-on-tall-screens

brianlovin committed:
Add comments to confusing code

brianlovin committed:
Merge pull request #2481 from withspectrum/fix-admin-shared

Make it possible for admin to import from shared/

spartDev committed:
create security middleware

spartDev committed:
apply security middleware to api

spartDev committed:
apply security middleware to hyperion

spartDev committed:
apply security middleware to plutot

littleStudent committed:
fix focus of the chatinput after a user composes a new DM

LetItRock committed:
Implement logging to Sentry; Adjust rate limiter values for Hyperion and API;

LetItRock commented:
Sorry for late update on that... I've adjusted values for rate limiter for API and Hyperion, also implemented logging to Sentry 🎉

dev-drprasad committed:
:bug: Remove invalid block types of draftjs in chat input

mxstbr committed:
Merge pull request #2912 from dev-drprasad/remove-list-items-markdown-plugin

:bug: Remove invalid block types of draftjs in chat input

mxstbr commented on hyperion/renderer/html-template.js:
```diff
@@ -72,7 +73,6 @@ export const getFooter = ({
       <script>window.__SERVER_STATE__=${serialize(state)}</script>
       <script>window.__DATA__=${serialize(data)}</script>
       <script defer="defer" type="text/javascript" src="https://cdn.polyfill.io/v2/polyfill.min.js"></script>
-      <script type="text/javascript" src="/static/js/bootstrap.js"></script>
```
This needs to be there, at least in production!

LetItRock replied:
ok, sorry I forgot to remove that...

mxstbr commented on index.html:
```diff
@@ -50,5 +51,5 @@
   </head>
   <body>
     <div id="root"></div>
-  <script type="text/javascript" src="/./static/js/bootstrap.js"></script><script type="text/javascript" src="/static/js/bundle.js"></script></body>
```
This needs to be there in production!

LetItRock replied:
sry

mxstbr committed:
Temporarily disable delete user integration test

This is the only one left that consistently fails and blocks CI, and I
have no clue how to fix it. All the others pass, so I think it makes
sense to skip this one (for now!) and fix it later.

mxstbr committed:
Fix circleci

spartDev committed:
remove middlewar on pluto

ryota-murakami committed:
fix(src/components/loading/style): align indent

LetItRock committed:
Revert removed bootstrap scripts

mxstbr reviewed (dismissed):
This looks great to me! :100: Thanks so much for digging into this.

/cc @brianlovin whaddaya think?

brianlovin committed:
Merge pull request #2915 from withspectrum/disable-delete-user-spec

Temporarily disable delete user integration test

brianlovin committed:
Merge pull request #2911 from littleStudent/focus-chatinput-after-composing-a-new-DM

fix focus of the chatinput after a user composes a new DM #2493

brianlovin committed:
Merge pull request #2890 from withspectrum/pagination-on-tall-screens

Pagination on tall screens

brianlovin committed:
Merge pull request #2916 from ryota-murakami/fix/remain-css-tab

fix(src/components/loading/style): align indent

brianlovin committed:
Small pronoun change

brianlovin commented:
This looks good to me overall; obviously we're shooting in the dark a bit regarding the limits and timeframes; @mxstbr would you be cool deploying this to alpha first and let's keep an eye on this for a while to make sure our limits feel good!

alejandronanez commented on api/routes/middlewares/index.js:
```diff
@@ -50,6 +50,12 @@ import passport from 'passport';
 
+// rate limiter middleware to reduce API calls to server within time frame
+import rateLimiter from 'shared/middlewares/rateLimiter';
+const THREE_HUNDRED = 300;
```
Hello @LetItRock,

What do you think about giving this a more _generic_ name, maybe something like `BOTTOM_LIMIT` and `TOP_LIMIT`? Imagine if you Spectrum decided to change `300` to `800`, then you'd have to change the variable name too.

brianlovin replied:
That's a good point; could be like:
`REQUESTS_ALLOWED` and `TIMEFRAME` - generic enough, and maps to what the arguments are doing in the function?

LetItRock replied:
sure I'll do that, you are both right its better to have here more generic name...

LetItRock committed:
More generic names for rate limiter constants

LetItRock committed:
Merge branch 'futures/node-api-requests-limiter' of https://github.com/LetItRock/spectrum into futures/node-api-requests-limiter

mxstbr committed:
Merge branch 'alpha' into securityMiddleware

mxstbr committed:
Add dependencies to ap

mxstbr committed:
Merge pull request #2909 from spartDev/securityMiddleware

Security middleware implementation

mxstbr commented:
Good call @brianlovin—I'm deploying these changes to `alpha.spectrum.chat` now!

mxstbr committed:
Merge branch 'futures/node-api-requests-limiter' of git://github.com/LetItRock/spectrum into LetItRock-futures/node-api-requests-limiter

mxstbr committed:
Add missing dependencies to api

mxstbr commented:
This was deployed to the API and will be out for hyperion shortly—let's keep an eye out on alpha and see if this breaks anything in the common users usecase.

brianlovin commented:
Closed with #4647 <3
